### ISSUE TITLE
force_ndarray is mutually exclusive with negative integer exponents

### ISSUE THREAD

**Issue by keewis** [hgrecco/pint#1295]
It seems that due to the way `parse_expression` works, using negative exponents with `force_ndarray=True` results in an error:
```python
In [2]: ureg = pint.UnitRegistry()
   ...: ureg("m ** -1")
Out[2]: 1.0 <Unit('1 / meter')>

In [3]: ureg2 = pint.UnitRegistry(force_ndarray=True)
   ...: ureg2("m ** -1")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-e15cfb26c53e> in <module>
      1 ureg2 = pint.UnitRegistry(force_ndarray=True)
----> 2 ureg2("m ** -1")

.../lib/python3.9/site-packages/pint/registry.py in parse_expression(self, input_string, case_sensitive, use_decimal, **values)
   1261         gen = tokenizer(input_string)
   1262 
-> 1263         return build_eval_tree(gen).evaluate(
   1264             lambda x: self._eval_token(x, case_sensitive=case_sensitive, **values)
   1265         )

.../lib/python3.9/site-packages/pint/pint_eval.py in evaluate(self, define_op, bin_op, un_op)
     91                 raise DefinitionSyntaxError('missing binary operator "%s"' % op_text)
     92             left = self.left.evaluate(define_op, bin_op, un_op)
---> 93             return bin_op[op_text](left, self.right.evaluate(define_op, bin_op, un_op))
     94         elif self.operator:
     95             # unary operator

.../lib/python3.9/site-packages/pint/quantity.py in wrapped(self, *args, **kwargs)
    113         elif isinstance(other, list) and other and isinstance(other[0], type(self)):
    114             return NotImplemented
--> 115         return f(self, *args, **kwargs)
    116 
    117     return wrapped

.../lib/python3.9/site-packages/pint/quantity.py in __pow__(self, other)
   1494                     units = new_self._units ** exponent
   1495 
-> 1496             magnitude = new_self._magnitude ** exponent
   1497             return self.__class__(magnitude, units)
   1498 

ValueError: Integers to negative integer powers are not allowed.
```
To be clear, that's a `numpy` issue: `np.array(1) ** -1` will raise the same error.

Not sure what to do about this: I would like to get this fixed somehow but this seems tricky (we seem to just evaluate the parsed syntax tree).

xref xarray-contrib/cf-xarray#197

cc @jthielen

**Comment by cpascual** [comment]
The error makes sense to me in the context of numpy **integer** arrays. Note that `np.array(1, dtype='float') ** -1` does work. The behaviour just becomes surprising in pint because pint blurs the distinction between floats and ints.

Without knowing the details of the underlying code, I wonder if it would be possible to replace the following code in  pint.quantity::__pow__():

```python
magnitude = new_self._magnitude ** exponent
```
by somethis like:
```python
try:
    magnitude = new_self._magnitude ** exponent
except ValueError:
    magnitude = new_self._magnitude.astype(float) ** exponent
```

**Comment by keewis** [comment]
I would like to avoid changing the behavior of the `Quantity` class: it should be a duck array and not deviate from `numpy`. This:
```python
ureg.Quantity(np.array([1, 2], dtype=int), "m") ** -1
```
should still raise. However, having it raise when parsing the expression seems like a implementation bug:
```python
ureg.parse_expression("2.0 * m ** -1")
```
is actually
```python
ureg.parse_expression("2.0 * (1 * m) ** -1")
```
if possible, I would like to make the evaluation of the parsed expression a bit smarter and cast `1 * m` to `float` if the exponent is negative.

For reference, the main reason for trying to fix this is getting the CF units to work with `force_ndarray`: `ureg("m s-2")`

**Comment by jthielen** [comment]
I definitely agree with @keewis with regards to the `Quantity` class behavior. To give another view, how I see it is that `parse_expression` should be "smarter" in either its choice of multiplicative identity (only using the integer `1` gives issues) or expression parsing order (so that the "unit part" of `"2.0 * m ** -1"` can be recognized before attaching a numeric value to it)

**Comment by hgrecco** [comment]
I would argue that when the user puts `"2.0 * m ** -1"` they always mean `"2.0 * (m ** -1)"`. As @jthielen was mentioning, I wonder if we can tweak the associativy rules of the parser to deal with this.

**Comment by keewis** [comment]
I didn't think this through, but another option could be to only convert to a `Quantity` when necessary: `parse_expression` would work with units until it encounters a number / constant or after computing the final result. Assuming that the test coverage is high enough, that fix would be surprisingly easy (review is still necessary):
```diff
diff --git a/pint/registry.py b/pint/registry.py
index 6b2697f..099bbab 100644
--- a/pint/registry.py
+++ b/pint/registry.py
@@ -1149,16 +1149,13 @@ class BaseRegistry(metaclass=RegistryMeta):
         token_text = token[1]
         if token_type == NAME:
             if token_text == "dimensionless":
-                return 1 * self.dimensionless
+                return self.dimensionless
             elif token_text in values:
-                return self.Quantity(values[token_text])
+                return values[token_text]
             else:
-                return self.Quantity(
-                    1,
-                    self.UnitsContainer(
-                        {self.get_name(token_text, case_sensitive=case_sensitive): 1}
-                    ),
-                )
+                return self.Unit(self.UnitsContainer(
+                    {self.get_name(token_text, case_sensitive=case_sensitive): 1}
+                ))
         elif token_type == NUMBER:
             return ParserHelper.eval_token(token, non_int_type=self.non_int_type)
         else:
@@ -1260,9 +1257,13 @@ class BaseRegistry(metaclass=RegistryMeta):
         input_string = string_preprocessor(input_string)
         gen = tokenizer(input_string)
 
-        return build_eval_tree(gen).evaluate(
+        result = build_eval_tree(gen).evaluate(
             lambda x: self._eval_token(x, case_sensitive=case_sensitive, **values)
         )
+        if isinstance(result, self.Unit):
+            result = self.Quantity(1, result)
+
+        return result
 
     __call__ = parse_expression
```
this won't support `(1 * s) ** -1`, but I guess that's a user error?

**Comment by hgrecco** [comment]
@keewis interesting finding! I think you are in the track. The current state of that function (`_eval_token`) comes from the past but I think you are proposing the right thing: "only return a quantity at the end"

However, I guess your change will break the following `ureg("meter + meter")`, right?

This `(1 * s) ** -1` this is indeed a user error and should be treated like that.

**Comment by keewis** [comment]
> However, I guess your change will break the following `ureg("meter + meter")`, right?

indeed. I'm not sure how to work around that: we could special-case `operator.add` / `operator.sub` in `EvalTree.evaluate` but then `ureg.parse_expression("(m + m) ** -1")` would fail so we'd also have to special-case `operator.pow` (which makes the "use units as long as possible" obsolete). Also, if that's something `pint` should support it should we should add a test case for it.

Somewhat related: is there a reason why `ureg.__call__` is a alias of `parse_expression` and not `parse_units`? I think `parse_expression` is not as useful in combination with `numpy` as it is with standard python types (`int`, `float`, `decimal`, `fraction`), or `uncertainties` (?)

**Comment by hgrecco** [comment]
I see two options:
1. improve the parser so that when there is a sum of two units, a quantity is returned (i.e. some kind of casting)
2. implement `__add__` and `_sub__` for `Unit` class

We can discuss `__call__` in another separate issue as changing this would be a breaking change.

**Comment by hgrecco** [comment]
The first option can be done quite easily I think. In addition to the change proposes by @keewis, in pint_eval we should doe

```python

def _op_unit(op)
    def _inner(left, right):
        if isintance(left, Unit) and isinstance(left, Unit):
            return  op(1 * left, 1 * right)
        return op(left, right)
    return _inner

_BINARY_OPERATOR_MAP["+"] = _op_unit(operator.add)
_BINARY_OPERATOR_MAP["-"] = _op_unit(operator.sub)
```

**Comment by keewis** [comment]
wouldn't it be easier to just override `**`? When I suggested to use units as long as possible I didn't see `_BINARY_OPERATOR_MAP`:
```python
def _power(left, right):
    if isinstance(left, Quantity) and is_duck_array(left) and left.dtype.kind not in "cf" and right < 0:
        left = left.astype(float)
    return operator.pow(left, right)

_BINARY_OPERATOR_MAP["**"] = _power
```

**Comment by hgrecco** [comment]
I think your solution should work, but I wonder if your other solution is more appropiate. To be honest, I am not sure.

**Comment by keewis** [comment]
actually, I'm not sure I understand the point of supporting `"m + m"`: this is currently equivalent to `"1 m + 1 m"` and returns `2 * ureg.m` but I don't think it should (it might make more sense to have it return `ureg.m` or raise an error).

If I'm not missing anything we might have a few options:
1. override `"**"` to cast integer arrays to float
2. use units as long as possible (and deprecate `"m + m"`)
3. use units as long as possible (but keep supporting `"m + m"`, involves overriding `"+"`, `"-"`, and `"**"`)
4. switch `ureg.__call__` to `parse_units` (I'll open a new issue to discuss this separately) and document that `ureg.parse_expression` fails with `force_ndarray` / `force_ndarray_like`, `int` dtypes and negative exponents

and combinations, like 1 + 2 (support something like `"(1 m + 1 m) ** -1"` but not `"(m + m) ** -1"`)

**Comment by cpascual** [comment]
I agree in that  `"(m + m )"` returning `2 * ureg.m` doesn't make sense, so I like option 1+2 (for simplicity and correctness)
